shell/platform/android/io/flutter/embedding/android/FlutterActivityAndFragmentDelegate.java

```diff
@@ -898,6 +906,28 @@ void onUserLeaveHint() {
     }
   }
 
+  /**
+   * Invoke this from {@code Activity#onWindowFocusChanged()}.
+   *
+   * <p>A {@code Fragment} host must have its containing {@code Activity} forward this call so that
+   * the {@code Fragment} can then invoke this method.
+   */
+  void onWindowFocusChanged(boolean hasFocus) {
+    ensureAlive();
+    Log.v(TAG, "Received onWindowFocusChanged: " + (hasFocus ? "true" : "false"));
+    if (host.shouldDispatchAppLifecycleState() && flutterEngine != null) {
+      // This may notify in a redundant fashion with onResume and onPause. The
+      // lifecycle channel will deduplicate, but since the Android documentation
+      // says that we shouldn't rely on any particular order between lifecycle
+      // events and onWindowFocusChanged, we notify in both cases.
+      if (hasFocus) {
+        flutterEngine.getLifecycleChannel().appIsResumed();
```

Comment:
From that disclaimer in the [`onWindowFocusChanged`](https://developer.android.com/reference/android/app/Activity#onWindowFocusChanged(boolean)) doc about "any particular order", it seems possible that we could get `onWindowFocusChanged(true)` before we get `onResume()`, while we're still in paused state.

If that does happen, it'd probably be best to hold off on `appIsResumed`. That way we preserve an invariant that when we say `AppLifecycleState.resumed` on Android, it always means at least that the app or Flutter host view is in a resumed state.

Reply:
This comment might indicate that we need to keep track on onPaused and only hold the resumed state onWindowFocusedChangedI(true) if we got an onPause. 

What makes this difficult to reason about it it is a path where many customers will have an assumption that is  unlikely to be tested and those together can lead to really frustrated users.

Reply:
The thing I'm worried about is having more than one source for the state Flutter thinks the app is in.  If I keep information in this class, or in the `LifecycleChannel` class, so that I know if the activity is already resumed, or in a particular state, then also hold the state in the Dart code in the framework, then it's possible for them to get out of sync. If I send all the state changes to the framework and deduplicate there, then I only have one place where the state is stored, and it can't be out of sync.

Reply:
That's not entirely true though. The Android framework also defines a source-of-truth for both Android lifecycle state and window focus state, which could already conceivably be out of sync with the Dart state. Unfortunately, vanilla Android does not expose either of these states as directly introspectable (AndroidX provides an API for introspecting the lifecycle state). It doesn't seem to me like tracking the Android state locally is any worse, unless the framework majorly violates its own contracts.

Reply:
Okay, good point.  No, the framework just takes what it is told on the `flutter/lifecycle` channel as gospel.  It doesn't conditionally do anything else. If I plugin were to talk to it on that channel, it would be confused, but I supposed you get what you ask for there.